harness/determined/_checkpoint_manager.py

```diff
@@ -0,0 +1,122 @@
+        ckpt = {
+            "trial_id": self.trial_id,
+            "batch_number": self.batch_number,
+            "start_time": self.start_time.astimezone(timezone.utc).strftime(
+                "%Y-%m-%dT%H:%M:%S.%fZ"
+            ),
+            "state": self.state,
+            "experiment_config": self.experiment_config,
+            "experiment_id": self.experiment_id,
+            "hparams": self.hparams,
+        }
+
+        r = req.post(self.master_url, "/api/v1/checkpoints", body={"checkpoint": ckpt})
+        if not hasattr(r, "headers"):
+            raise Exception(r)
+        return self
+
+    def complete(
+        self,
+        uuid: str,
+        resources: Optional[Dict[str, Any]] = None,
+        metadata: Optional[Dict[str, Any]] = None,
+        framework: Optional[str] = None,
+        format: Optional[str] = None,
+    ) -> None:
+        logging.info(
+            f"Finishing saving checkpoint {uuid} for "
+            f"trial {self.trial_id} batch {self.batch_number}."
+        )
+        start_time = self.start_time.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%fZ")
+        end_time = (
+            datetime.now(timezone.utc).astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%fZ")
+        )
+        ckpt = {
+            "uuid": uuid,
+            "trial_id": self.trial_id,
+            "batch_number": self.batch_number,
+            "start_time": start_time,
+            "end_time": end_time,
+            "state": "STATE_COMPLETED",
+            "experiment_config": self.experiment_config,
+            "experiment_id": self.experiment_id,
+            "hparams": self.hparams,
+            "resources": resources,
+            "metadata": metadata,
+            "determined_version": det.__version__,
+            "framework": framework,
+            "format": format,
+        }
+
+        r = req.put(self.master_url, "/api/v1/checkpoints", body={"checkpoint": ckpt})
+        if not hasattr(r, "headers"):
+            raise Exception(r)
+
+        self.state = "STATE_COMPLETED"
+
+    def failed(self) -> None:
+        logging.info(
+            f"Failed saving checkpoint for trial {self.trial_id} batch {self.batch_number}."
+        )
+        start_time = self.start_time.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%fZ")
+        end_time = (
+            datetime.now(timezone.utc).astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%fZ")
+        )
+        ckpt = {
+            "trial_id": self.trial_id,
+            "batch_number": self.batch_number,
+            "start_time": start_time,
+            "end_time": end_time,
+            "state": "STATE_ERRORED",
+            "experiment_config": self.experiment_config,
+            "experiment_id": self.experiment_id,
+            "hparams": self.hparams,
+            "determined_version": det.__version__,
+            "format": format,
+        }
+
+        r = req.put(self.master_url, "/api/v1/checkpoints", body={"checkpoint": ckpt})
```

Comment:
Should we maybe be consuming auto generated bindings for this API since they could exist?

Reply:
yes, we should. However, we don't have generated bindings now and this PR doesn't aim at changing anything about it.

Reply:
Yeah, I know we're starting to consume the API all over the place.. we should get that figured out soon. I'll put it on th infra-ag agenda.